presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoDriver.java

```diff
@@ -57,31 +80,261 @@
+     *       releases read lock
+     *
+     * The last case of interest is the process that is using PrestoDriver
+     * calling prestoDriver.close(). This invalidates all of the cached
+     * QueryExecutors regardless of their refcount, and in doing so closes
+     * their HTTP clients. If anybody is still trying to use a
+     * PrestoConnection created by the driver, they're SOL.
+     */
     public PrestoDriver()
     {
-        this.queryExecutor = QueryExecutor.create(DRIVER_NAME + "/" + DRIVER_VERSION);
+        this.jettyIoPool = new JettyIoPool("presto-jdbc", new JettyIoPoolConfig());
+        this.cacheLock = new ReentrantReadWriteLock();
+        this.queryExecutorCleanupService = new ScheduledThreadPoolExecutor(1);
+
+        this.queryExecutorCache = CacheBuilder.newBuilder()
+                .removalListener(new RemovalListener<Map<Object, Object>, QueryExecutorHolder>() {
+                        @Override
+                        public void onRemoval(RemovalNotification<Map<Object, Object>, QueryExecutorHolder> notification)
+                        {
+                            notification.getValue().closeQueryExecutor();
+                        }
+                    })
+                .build(new CacheLoader<Map<Object, Object>, QueryExecutorHolder>() {
+                        @Override
+                        public QueryExecutorHolder load(Map<Object, Object> clientProperties)
+                        {
+                            return new QueryExecutorHolder(
+                                    QueryExecutor.create(DRIVER_NAME + "/" + DRIVER_VERSION, jettyIoPool),
+                                    clientProperties);
+                        }
+                    });
     }
 
     @Override
     public void close()
     {
-        queryExecutor.close();
+        if (closed.compareAndSet(false, true)) {
+            queryExecutorCleanupService.shutdown();
+            queryExecutorCache.invalidateAll();
+            jettyIoPool.close();
+        }
+    }
+
+    private class QueryExecutorHolder
+    {
+        private QueryExecutor queryExecutor;
+
+        private Map<Object, Object> clientProperties;
+
+        private AtomicInteger refcount = new AtomicInteger(0);
+
+        private QueryExecutorHolder(QueryExecutor queryExecutor, Map<Object, Object> clientProperties)
+        {
+            this.queryExecutor = requireNonNull(queryExecutor, "wrapped is null");
+            this.clientProperties = requireNonNull(clientProperties, "clientProperties is null");
+        }
+
+        private void closeQueryExecutor()
+        {
+            queryExecutor.close();
+        }
+
+        public void release()
+        {
+            Runnable deferredClose = () -> {
+                cacheLock.writeLock().lock();
+                try {
+                    if (refcount.decrementAndGet() == 0) {
+                        queryExecutorCache.invalidate(clientProperties);
```

Comment:
I would rather move it to `PrestoConnection#close`, and leave here only `acquire`, `release` methods. So you don't need to hold `clientProperties` in this object, that are seems to do not belong here. You can make `release` to return `boolean` value, which indicates whether the `QueryExecutor` must be released.